### PR TITLE
MC 1.17 3D Biomes

### PR DESCRIPTION
**Pull request by NotStirred**
Seems to work correctly to me, though I'd like to test it with corgi's world gen fork PR before a merge

### CODE REVIEW

**Comment by NotStirred** [comment]
After testing with corgi's branch this works properly. Ready for merge

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/BigCube.java [hunk 1]
```diff
@@ -152,12 +148,12 @@ public BigCube(Level worldIn, CubePos cubePosIn, CubeBiomeContainer biomeContain
 //        this.gatherCapabilities();
     }
 
-    public BigCube(Level worldIn, CubePrimer cubePrimerIn) {
+    public BigCube(Level worldIn, CubePrimer cubePrimerIn, @Nullable Consumer<BigCube> consumer) {
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/BigCube.java]: This consumer should have a better name

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubePrimerWrapper.java [hunk 2]
```diff
@@ -135,9 +135,9 @@ public BigCube getCube()
         return this.cube.isEmptyCube();
     }
 
-    @Deprecated @Override public ChunkBiomeContainer getBiomes() { return this.getCubeBiomes(); }
+    @Deprecated @Override public ChunkBiomeContainer getBiomes() { throw new UnsupportedOperationException("Chunk method called on a cube"); }
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubePrimerWrapper.java]: Redirect anyway and let the `BigCube.getBiomes()` throw

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java [hunk 3]
```diff
@@ -1,13 +1,74 @@
...
+    @Shadow private RegistryAccess registryAccess;
+
+    @Shadow public abstract ClientLevel getLevel();
+
+//    private static final ChunkBiomeContainer dummyContainer = new ChunkBiomeContainer(new IdMapper<>(), new Biome[ChunkBiomeContainer.BIOMES_SIZE]);
+//    @Nullable
+//    @Redirect(method = "handleLevelChunk", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/multiplayer/ClientChunkCache;replaceWithPacketData(IILnet/minecraft/world/level/chunk/ChunkBiomeContainer;Lnet/minecraft/network/FriendlyByteBuf;Lnet/minecraft/nbt/CompoundTag;I)Lnet/minecraft/world/level/chunk/LevelChunk;"))
+//    private LevelChunk on$handleLevelChunk(ClientChunkCache clientChunkCache, int i, int j, ChunkBiomeContainer chunkBiomeContainer, FriendlyByteBuf friendlyByteBuf, CompoundTag compoundTag, int k) {
+//        return clientChunkCache.replaceWithPacketData(i, j, dummyContainer, friendlyByteBuf, compoundTag, k);
+//    }
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java]: Remove that comment or fix the code

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java [hunk 4]
```diff
@@ -1,13 +1,74 @@
...
+    public void handleLevelChunk(ClientboundLevelChunkPacket clientboundLevelChunkPacket, CallbackInfo ci) {
+        //TODO: implement non-cubic world, inject at head and cancel
+        ci.cancel();
+
+        PacketUtils.ensureRunningOnSameThread(clientboundLevelChunkPacket, (ClientPacketListener)(Object)this, this.minecraft);
+        int chunkX = clientboundLevelChunkPacket.getX();
+        int chunkZ = clientboundLevelChunkPacket.getZ();
+
+        //For a cc world we will always get null biomes
+//        ColumnBiomeContainer biomeContainer = new ColumnBiomeContainer(this.registryAccess.registryOrThrow(Registry.BIOME_REGISTRY), new int[ColumnBiomeContainer.BIOMES_SIZE]);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java]: Why is that commented out line here

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java [hunk 5]
```diff
@@ -1,13 +1,74 @@
...
+        @SuppressWarnings("ConstantConditions") //Not an NPE because this method is client-side.
+        ColumnBiomeContainer biomeContainer = new ColumnBiomeContainer(minecraft.level.registryAccess().registryOrThrow(Registry.BIOME_REGISTRY), new int[ColumnBiomeContainer.BIOMES_SIZE]);
+        biomeContainer.setLevel(minecraft.level);
+        LevelChunk levelChunk = this.level.getChunkSource().replaceWithPacketData(chunkX, chunkZ, biomeContainer, clientboundLevelChunkPacket.getReadBuffer(), clientboundLevelChunkPacket.getHeightmaps(), clientboundLevelChunkPacket.getAvailableSections());
+
+        for (int k = this.level.getMinSection(); k < this.level.getMinSection() - 1; ++k) {
+            this.level.setSectionDirtyWithNeighbors(chunkX, k, chunkZ);
+        }
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java]: Delete that loop

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java [hunk 6]
```diff
@@ -1,13 +1,74 @@
...
+
+        if (levelChunk != null) {
+            for (CompoundTag compoundTag : clientboundLevelChunkPacket.getBlockEntitiesTags()) {
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinClientPacketListener.java]: We are not sending block entities in this packet

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinRenderChunk.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    @Shadow protected abstract double getDistToPlayerSqr();
+
+    @Shadow protected abstract boolean doesChunkExistAt(BlockPos blockPos);
+
+    @Shadow @Final private BlockPos.MutableBlockPos[] relativeOrigins;
+
+    @SuppressWarnings("target") @Shadow(aliases = "field_20833", remap = false) ChunkRenderDispatcher this$0;
+
+    @Inject(method = "doesChunkExistAt", at = @At("HEAD"), cancellable = true)
+    private void on$doesChunkExists(BlockPos blockPos, CallbackInfoReturnable<Boolean> cir) {
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinRenderChunk.java]: Why are you using `on$` in method names?

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinRenderChunk.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        cir.setReturnValue(((ICubicWorld) ((ChunkRenderDispatcherAccess) this$0).getLevel()).getCube(
+            Coords.blockToCube(blockPos.getX()),
+            Coords.blockToCube(blockPos.getY()),
+            Coords.blockToCube(blockPos.getZ()),
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/MixinRenderChunk.java]: Also check for column

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/MixinServerChunkProvider.java [hunk 9]
```diff
@@ -261,7 +262,7 @@ public void blockChanged(BlockPos pos) {
     }
 
     @Inject(method = "tickChunks",
-            at = @At(value = "INVOKE", target = "Lnet/minecraft/server/level/ChunkMap;getChunks()Ljava/lang/Iterable;"))
+            at = @At(value = "INVOKE", target = "Lnet/minecraft/server/level/ChunkMap;getChunks()Ljava/lang/Iterable;"), locals = LocalCapture.CAPTURE_FAILHARD)
     private void tickSections(CallbackInfo ci) {
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/MixinServerChunkProvider.java]: Remove the unnecessary local capture

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java [hunk 10]
```diff
@@ -42,10 +46,20 @@
...
     @Shadow @Final private Map<BlockPos, CompoundTag> pendingBlockEntities;
 
+    @Shadow private ChunkBiomeContainer biomes;
+
     @Override public boolean isYSpaceEmpty(int startY, int endY) {
         return false;
     }
 
+    @Inject(method = "<init>(Lnet/minecraft/world/level/Level;Lnet/minecraft/world/level/ChunkPos;Lnet/minecraft/world/level/chunk/ChunkBiomeContainer;Lnet/minecraft/world/level/chunk/UpgradeData;Lnet/minecraft/world/level/TickList;Lnet/minecraft/world/level/TickList;J[Lnet/minecraft/world/level/chunk/LevelChunkSection;Ljava/util/function/Consumer;)V", at = @At("RETURN"))
+    private void on$init(Level level, ChunkPos chunkPos, ChunkBiomeContainer chunkBiomeContainer, UpgradeData upgradeData, TickList<Block> tickList, TickList<Fluid> tickList2, long l, LevelChunkSection[] levelChunkSections, Consumer<LevelChunk> consumer, CallbackInfo ci) {
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java]: Again, don't use `$` in method names unless necessary

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java [hunk 11]
```diff
@@ -42,10 +46,20 @@
...
+        //TODO: is cubicworld
+        if(!(biomes instanceof ColumnBiomeContainer)) //Client will already supply a ColumnBiomeContainer, server will not
+            this.biomes = new ColumnBiomeContainer(level.registryAccess().registryOrThrow(Registry.BIOME_REGISTRY), new Biome[ColumnBiomeContainer.BIOMES_SIZE]);
+        ((ColumnBiomeContainer) biomes).setLevel(level);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java]: Why didn't you include level in the constructor for ColumnBiomeContainer?

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java [hunk 12]
```diff
@@ -180,4 +194,10 @@ private boolean getLoadedFromBlockEntity(LevelChunk chunk, BlockEntity tileEntit
     private boolean getLoadedFromBlockPos(LevelChunk chunk, BlockPos pos) {
         return ((BigCube)this.getCube(Coords.blockToSection(pos.getY()))).isInLevel();
     }
+
+    @Inject(method = "getBiomes", at = @At("HEAD"), cancellable = true)
+    public void getBiomes(CallbackInfoReturnable<ChunkBiomeContainer> cir) {
+        //TODO: IF cubicworld don't cancel
+        cir.setReturnValue(this.biomes);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java]: vanilla does the same, unnecessary inject

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/biome/ColumnBiomeContainer.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    private static final Biome DUMMY_BIOME = BuiltinRegistries.BIOME.get(Biomes.FOREST.location());
+
+    private Level level;
+
+    public ColumnBiomeContainer(IdMap<Biome> idMap, Biome[] biomes) {
+        super(idMap, biomes);
+    }
+
+    public ColumnBiomeContainer(IdMap<Biome> idMap, int[] is) {
+        super(idMap, is);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/biome/ColumnBiomeContainer.java]: You don't need an id or biome array, just pass in array of length 0 to super()

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinProtoChunk.java [hunk 14]
```diff
@@ -1,13 +1,29 @@
...
+import java.util.concurrent.CompletionException;
+
 @Mixin(ProtoChunk.class)
-public class MixinProtoChunk {
+public abstract class MixinProtoChunk {
+    @Shadow @Final private ChunkPos chunkPos;
+
+    @Shadow public abstract ChunkStatus getStatus();
+
+    @Shadow @Final private LevelHeightAccessor levelHeightAccessor;
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinProtoChunk.java]: Are you using these anywhere?

File: src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/network/MixinClientboundLevelChunkPacket.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+@Mixin(ClientboundLevelChunkPacket.class)
+public class MixinClientboundLevelChunkPacket {
+
+    /**
+     * @author NotStirred
+     * @reason CC doesn't need to send chunk biome data, as it's in cubes.
+     */
+    @Redirect(method = "<init>(Lnet/minecraft/world/level/chunk/LevelChunk;)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/level/chunk/ChunkBiomeContainer;writeBiomes()[I"))
+    public int[] on$writeBiomes(ChunkBiomeContainer chunkBiomeContainer) {
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/network/MixinClientboundLevelChunkPacket.java]: No `$` in method names

File: src/main/java/io/github/opencubicchunks/cubicchunks/network/PacketCubes.java [hunk 16]
```diff
@@ -53,6 +64,14 @@ public PacketCubes(List<BigCube> cubes) {
         // one long stores information about 64 chunks
         int length = MathUtil.ceilDiv(cubes.length, 64);
         this.cubeExists = BitSet.valueOf(buf.readLongArray(new long[length]));
+
+        ByteBuffer biomeBuffer = ByteBuffer.allocate(count * CubeBiomeContainer.CUBE_BIOMES_SIZE * 4);
+        buf.readBytes(biomeBuffer.array());
+        IntBuffer intBuf = biomeBuffer.asIntBuffer();
+
+        biomes = new int[intBuf.remaining()];
+        intBuf.get(biomes);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/network/PacketCubes.java]: use read/writeVarIntArray

File: src/main/java/io/github/opencubicchunks/cubicchunks/network/PacketCubes.java [hunk 17]
```diff
@@ -99,8 +123,14 @@ public static void handle(PacketCubes packet, Level worldIn) {
                 int y = pos.getY();
                 int z = pos.getZ();
 
+
+                int[] cubeBiomes = new int[CubeBiomeContainer.CUBE_BIOMES_SIZE];
+                if (cubeBiomes.length >= 0)
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/network/PacketCubes.java]: why?

File: src/main/java/io/github/opencubicchunks/cubicchunks/utils/Coords.java [hunk 18]
```diff
@@ -59,6 +59,7 @@
     private static final int INDEX_TO_N_Y = LOG2_BLOCK_SIZE - 4;
     private static final int INDEX_TO_N_Z = INDEX_TO_N_Y * 2;
 
+    private static final int BLOCK_TO_BIOME_SHIFT = MathUtil.log2(IBigCube.DIAMETER_IN_BLOCKS) - (MathUtil.log2(IBigCube.DIAMETER_IN_BLOCKS) - 2);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/utils/Coords.java]: that's just `2`

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]

**Pull request merged**